### DISCUSSION TITLE
Support for emacsclient/daemon mode

### DISCUSSION THREAD

**Discussion by bobbyrullo** [jamescherti/easysession.el#15] [Ideas]
Hi there! I run emacs in daemon mode. If you use 'after-init-hook as specified in the installation, it crashes on startup. What I did instead which works is to load easysession when emacsclient starts, using the focus-in-hook:

```
(defun bobby/easy-session ()
  (easysession-load-including-geometry)
  (easysession-save-mode)
  (remove-hook 'focus-in-hook #'bobby/easy-session))

;;https://github.com/jamescherti/easysession.el
(use-package easysession
  :ensure t
  :custom
  (easysession-save-interval (* 10 60))
  :init
  (add-hook 'focus-in-hook #'bobby/easy-session))
```

I'm not sure if there is a better way, but would be nice to add to docs or something so that emacsclient folks have an easier time.

**Comment by jamescherti** [comment]
Thank you for sharing your solution, @bobbyrullo! 

I was unable to reproduce the bug in Emacs 30. It may be specific to the version of Emacs you are using.

Running Emacs in daemon mode does require handling hooks differently, and your use of `focus-in-hook` (obsolete since Emacs 27.1; replaced with `after-focus-change-function`) to trigger loading the session makes sense.

One alternative you could consider is using `server-after-make-frame-hook` instead of `focus-in-hook`. 

Could you please try it? Here’s an example:
```elisp
(defun bobby/easy-session ()
  (easysession-load-including-geometry)
  (easysession-save-mode)
  (remove-hook 'server-after-make-frame-hook #'bobby/easy-session))

(use-package easysession
  :ensure t
  :custom
  (easysession-save-interval (* 10 60))
  :init
  (add-hook 'server-after-make-frame-hook #'bobby/easy-session))
```

**Question 1:** Which solution works better for you, `server-after-make-frame-hook` or `focus-in-hook`?
**Question 2:** Does `emacs-startup-hook` make Emacs server crash like `after-init-hook`: 
```elisp
(add-hook 'emacs-startup-hook #'easysession-load-including-geometry 98)
(add-hook 'emacs-startup-hook #'easysession-save-mode 99))
```
**Question 3:** What Emacs version are you using?

I plan to update the README.md after your confirmation.

**Reply by bobbyrullo** [reply]
> Question 1: Which solution works better for you, server-after-make-frame-hook or focus-in-hook?

They both work the same for me; I'll stick with server-after-make-frame-hook because it's not deprecated :) 

EDIT: Went back to my original solution; after all my buffers loaded and some other stuff happened (eg. lsp starting a bunch of gopls processes) it would crash the client :( 

> Question 2: Does emacs-startup-hook make Emacs server crash like after-init-hook:
Not exactly....the daemon started fine, but when I tried to connect via emacsclient it just hung for a very long time, no frames appeared, so I killed it.

> (add-hook 'emacs-startup-hook #'easysession-load-including-geometry 98)
> (add-hook 'emacs-startup-hook #'easysession-save-mode 99))
> Question 3: What Emacs version are you using?

29.1. The other thing is that I'm doing all this via x2GoClient; for some reason when I upgraded to 30 it would crash when I connected via x2GoClient. No idea if that's an x2Go thing, or an emacs thing or something specific to my box, but I didn't investigate very much since this is my work devbox and just wanted to  work :)
> 
>

**Reply by jamescherti** [reply]
Thank you for confirming. I will add this to the `README.md` file.
